ai/priority_algorithm/mutations.py

```diff
@@ -30,3 +38,186 @@ def mutate_random_swap(priority_team_set: PriorityTeamSet) -> PriorityTeamSet:
+                    ]
+                )
+            else:
+                satisfied_team = random.choice(satisfied_teams)
+                unsatisfied_team = random.choice(unsatisfied_teams)
+
+            # Perform the local max portion of the robinhood mutation
+            (
+                local_best_team_set,
+                local_best_team_set_score,
+            ) = perform_local_max_portion_of_robinhood(
+                cloned_priority_team_set,
+                priorities,
+                student_dict,
+                satisfied_team,
+                unsatisfied_team,
+            )
+
+            # Update the best team set if the local best team set is better
+            if local_best_team_set_score > best_team_set_score:
+                best_team_set = local_best_team_set
+                best_team_set_score = local_best_team_set_score
+    except ValueError:
+        return best_team_set
+    return best_team_set
+
+
+def perform_local_max_portion_of_robinhood(
+    cloned_priority_team_set: PriorityTeamSet,
+    priorities: List[Priority],
+    student_dict: Dict[int, Student],
+    selected_team_a: PriorityTeam,
+    selected_team_b: PriorityTeam,
+) -> (PriorityTeamSet, float):
+    """
+    Performs the local max portion of the robinhood mutation. This is the part where we generate all possible teams using the students from the two teams, and choose the best team.
+
+    returns (best_team_set, best_team_set_score)
+    """
+
+    # Init best team set
+    best_team_set: PriorityTeamSet = cloned_priority_team_set
+    best_team_set_score: float = cloned_priority_team_set.calculate_score(
+        priorities, student_dict
+    )
+
+    # List of all students in the two teams
+    students: List[int] = selected_team_a.student_ids + selected_team_b.student_ids
+
+    # Generate all possible teams using the students from the two teams
+    # Elements of this list are lists of student ids
+    possible_teams: List[List[int]] = list(
+        itertools.combinations(students, len(selected_team_b.student_ids))
+    )
+
+    # Calculate the score of each team
+    for team in possible_teams:
+        # Modify the cloned PriorityTeamSet to reflect the new team
+        selected_team_b.student_ids = team
+        selected_team_a.student_ids = [
+            student_id for student_id in students if student_id not in team
+        ]
+
+        # Calculate the score of the resulting PriorityTeamSet
+        cloned_priority_team_set.score = (
+            None  # Reset the score to force recalculation after modification
+        )
+        score = cloned_priority_team_set.calculate_score(priorities, student_dict)
+        if score > best_team_set_score:
+            # Clone the PriorityTeamSet because we will be modifying on the next iteration
+            best_team_set = cloned_priority_team_set.clone()
+            best_team_set_score = score
+
+    return best_team_set, best_team_set_score
+
+
+def valid_robinhood_arguments(
```

Comment:
this is cool but most of these cases are either checked before it gets here or don't matter, I'll leave some more comments but tldr is if the mutation would do nothing then just return the unedited teamset (so don't "mutate" it at all)